source/postprocess/geoid.cc

```diff
@@ -0,0 +1,975 @@
+
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace internal
+    {
+      /* Definitions for HarmonicCoefficients */
+
+      template <int dim>
+      HarmonicCoefficients<dim>::HarmonicCoefficients(const unsigned int max_degree)
+      {
+        // Allocate vectors of the appropriate size for 2D and 3D.  For 2D we
+        // need cylindrical harmonics, for 3D we need spherical harmonics.
+
+        unsigned int k= (dim == 2 ?
+                         2 * (max_degree + 1 ) :             //cylindrical harmonics
+                         (max_degree+1)*(max_degree+2)/2 ) ; //spherical harmonics
+        sine_coefficients.resize(k);
+        cosine_coefficients.resize(k);
+      }
+
+
+      /* Definitions for the MultipoleExpansion base class */
+
+      // Constructor.  Just need the initialize the coefficients struct.
+      template <int dim>
+      MultipoleExpansion<dim>::MultipoleExpansion(const unsigned int max_degree)
+        :
+        max_degree(max_degree),
+        coefficients(max_degree)
+      {}
+
+      template <int dim>
+      void
+      MultipoleExpansion<dim>::add_quadrature_point( const Point<dim> &, const double, const double )
+      {
+        AssertThrow(false, ExcNotImplemented() );
+      }
+
+      template <int dim>
+      double
+      MultipoleExpansion<dim>::evaluate( const Point<dim> &) const
+      {
+        AssertThrow(false, ExcNotImplemented() );
+        return 0.0;
+      }
+
+      //Return a reference to the internal representation of the multipole expansion
+      template <int dim>
+      const HarmonicCoefficients<dim> &
+      MultipoleExpansion<dim>::get_coefficients () const
+      {
+        return coefficients;
+      }
+
+      template <int dim>
+      unsigned int
+      MultipoleExpansion<dim>::degree () const
+      {
+        return max_degree;
+      }
+
+      //Zero out the expansion in 2D
+      template <>
+      void MultipoleExpansion<2>::clear()
+      {
+        for (unsigned int n = 2; n <= max_degree; ++n)
+          {
+            coefficients.sine_coefficients[n] = 0.0;
+            coefficients.cosine_coefficients[n] = 0.0;
+          }
+      }
+
+      //Zero out the expansion in 3D
+      template <>
+      void MultipoleExpansion<3>::clear()
+      {
+        for (unsigned int l = 2, k = 0; l <= max_degree; ++l)
+          for (unsigned int m = 0; m <= l; ++m, ++k)
+            {
+              coefficients.sine_coefficients[k] = 0.0;
+              coefficients.cosine_coefficients[k] = 0.0;
+            }
+      }
+
+      template <>
+      void MultipoleExpansion<2>::sadd( double s, double a, const MultipoleExpansion &M )
+      {
+        AssertThrow( coefficients.sine_coefficients.size() == M.get_coefficients().sine_coefficients.size() ,
+                     ExcInternalError() );
+
+        for (unsigned int n = 2; n <= max_degree; ++n)
+          {
+            coefficients.sine_coefficients[n] = s * coefficients.sine_coefficients[n] +
```

Comment:
Maybe I just did not understand this correctly, but in 2D you fill `sine_coefficients[2...end]`, but in 3D  you fill `sine_coefficients[0...end]`? Is this because of the different structure of the expansions?